crypto/src/test_utils/mod.rs

```diff
@@ -168,31 +174,153 @@ pub async fn run_test_with_deterministic_client_ids<const N: usize>(
     client_ids: [[&'static str; 3]; N],
     test: impl FnOnce([ClientContext; N]) -> std::pin::Pin<Box<dyn std::future::Future<Output = ()> + 'static>> + 'static,
 ) {
-    run_test_with_deterministic_client_ids_and_revocation(case, client_ids, &[], test).await
+    run_test_with_deterministic_client_ids_and_revocation(case, client_ids, [], &[], test).await
 }
 
-pub async fn run_test_with_deterministic_client_ids_and_revocation<const N: usize>(
+/// Generates 2 x509 test chains and federate them with each other
+pub fn init_federated_x509_test_chains<const N: usize, const F: usize>(
+    case: &TestCase,
+    client_ids: [[&'static str; 3]; N],
+    federated_client_ids: [[&'static str; 3]; F],
+    (params1, params2): (CertificateParams, CertificateParams),
+    revoked_display_names: &'static [&'static str],
+) -> (X509TestChain, X509TestChain) {
+    let mut chain1 = init_x509_test_chain(case, &client_ids, revoked_display_names, params1);
+    let mut chain2 = init_x509_test_chain(case, &federated_client_ids, revoked_display_names, params2);
+    chain1.cross_sign(&mut chain2);
+    (chain1, chain2)
+}
+
+pub async fn run_cross_signed_tests_with_client_ids<const N: usize, const F: usize>(
     case: TestCase,
     client_ids: [[&'static str; 3]; N],
+    federated_client_ids: [[&'static str; 3]; F],
+    (domain1, domain2): (&'static str, &'static str),
+    test: impl FnOnce([MlsCentral; N], [MlsCentral; F]) -> std::pin::Pin<Box<dyn std::future::Future<Output = ()> + 'static>>
+        + 'static,
+) {
+    assert!(case.is_x509(), "This is only supported for x509 test cases");
+    let _ = pretty_env_logger::try_init();
```

Comment:
Why not just use `pretty_env_logger::init()`, if we're discarding the return value?

Reply:
because it can panic if another one is set before

Reply:
Hm, can we ensure that `pretty_env_logger` is initialized only once, say, in some toplevel test fixture or something like that? I see it only being used in two places.

Reply:
Ah, good point. :+1: